### DISCUSSION TITLE
program list

### DISCUSSION THREAD

**Discussion by deopark** [pybricks/support#1353] [Q&A] [answered]
Download the custom(user) module and
Can I see the downloaded file on the hub?

Or can I download several programs? (like slots on the official app)

**Comment by ggramlich** [comment, marked as answer]
There is only one active program on the hub, but you can import functions from other files / modules, which will all be downloaded to the hub (from my understanding).

If you want to chose from multiple functions, take a look at the hub_menu from the tools
https://docs.pybricks.com/en/stable/tools/index.html#input-tools

**Reply by deopark** [reply]
Thank you! I learned how to import custom modules because of you.
Then, Can I get a list of custom modules(*.py) downloaded to the hub?

**Reply by dlech** [reply]
Only imported files/modules are downloaded to the hub, so you already have the list because it is the list of imports.

**Reply by dlech** [reply]
You should be able to get this list using `usys.modules`.

**Reply by BertLindeman** [reply]
I was looking for something like that, but ....
```python
import usys as sys
# help(sys)
from pybricks.tools import StopWatch, wait

# # Print the names of all imported modules
for module_name in sys.modules():
    print(module_name)
```

But that reports:
```
Traceback (most recent call last):
  File "show_modules.py", line 6, in <module>
AttributeError: 'module' object has no attribute 'modules'
```

Running: ```('technichub', '3.3.0', 'v1.20.0-23-g6c633a8dd on 2023-11-24')```

```python
help(sys)
```
reports:
```
object <module 'sys'> is of type module
  __name__ -- sys
  path -- ['', '.frozen']
  argv -- []
  version -- 3.4.0; Pybricks MicroPython v1.20.0-23-g6c633a8dd on 2023-11-24
  version_info -- (3, 4, 0)
  implementation -- ('micropython', (1, 20, 0), 'LEGO Technic Hub with STM32L431RC', 262)
  byteorder -- little
  stdin -- <io.FileIO 0>
  stdout -- <io.FileIO 1>
  stderr -- <io.FileIO 2>
  print_exception -- <function>
```

chatGPT suggests also:
```python
# Print the names of all imported modules using globals()
for module_name, module in globals().items():
    if isinstance(module, type(sys)):
        print(module_name)
```

**Reply by dlech** [reply]
Thanks for checking. It looks like we have the `usys.modules` feature currently disabled in `bricks/_common/mpconfigport.h`.

```c
#define MICROPY_PY_SYS_MODULES                  (0)
```

The ChatGPT suggestion sounds like a nice way to get the job done.

**Reply by dlech** [reply]
Riffing on that, something like this would get a dict similar to `sys.modules`.

```python
modules = {n: m for n, m in globals().items() if isinstance(m, type(sys))}
```

**Reply by deopark** [reply]
> Only imported files/modules are downloaded to the hub, so you already have the list because it is the list of imports.

Thank you for your reply.

Can I know which custom(user define) modules (user.py ) are downloaded? before I import.
Also, what if I want to delete a downloaded custom file (.py) or see if it exists?

**Reply by dlech** [reply]
> Can I know which custom(user define) modules (user.py ) are downloaded?

If there are any import statements like `import user` or `from user import ...` then `user.py` will be downloaded to the hub along with the current script that is importing it. All files are downloaded as a single package, so downloading and running a new program replaces all files previously downloaded to the hub.

**Reply by deopark** [reply]
Now understand!!
Thank you, Master!

**Comment by ggramlich** [comment]
I actually just saw an interesting use of "dynamic" imports in combination with hub_menu. I had never thought of using import this way.

https://youtu.be/xMVo1WgQSPk?si=vtkIwPERvchodVIF&t=2256

![image](https://github.com/pybricks/support/assets/69276/d0ceb9d3-de6e-4a06-9f6f-b63c21551e14)

**Comment by laurensvalk** [comment]
We've just added a tutorial that also answers this question. It works both with Python and block-based coding.

[![spike-hubmenu-yt](https://github.com/pybricks/support/assets/12326241/eeba4d24-4d68-40fd-8718-71e11e6634cf)](https://pybricks.com/project/spike-hub-menu/)

**Reply by derrickgw** [reply]
Would love to see an updated tutorial for this where two different menu options use the same motor or sensor.
